src/dos/dosirq.c

```diff
@@ -0,0 +1,386 @@
+	return irq;
+}
+
+void irq_unhook(struct irq_handle *irq)
+{
+	_go32_dpmi_seginfo info;
+
+	if (!irq)
+		return;
+
+	/* Restore the interrupt vector */
+	irq_disable(irq);
+	info.pm_offset = irq->prev_offset;
+	info.pm_selector = irq->prev_selector;
+	_go32_dpmi_set_protected_mode_interrupt_vector(irq->int_num, &info);
+
+	/* Unlock the interrupt handler */
+	dpmi_unlock_linear_region_base(irq->c_handler, irq->handler_size);
+
+	/* Unlock the irq_handle structure */
+	dpmi_unlock_linear_region_base(irq, sizeof(struct irq_handle));
+
+	info.pm_offset = irq->handler;
+	_free_iret_wrapper(&info);
+
+	/* If IRQ was enabled before we hooked, restore enabled state */
+	if (irq->pic_mask)
+		irq_enable(irq);
+	else
+		irq_disable(irq);
+
+	free(irq);
+}
+
+#elif defined(__WATCOMC__)
+
+struct irq_handle *irq_hook(int irqno, irq_handler handler, void (*end)())
+{
+	unsigned long size = (char *)end - (char near *)handler;
+	int intno = (irqno > 7) ? (irqno + 104) : (irqno + 8);
+	struct irq_handle *irq;
+	irq_handler old_vect;
+
+	old_vect = _dos_getvect(intno);
+
+	/* Lock the interrupt handler in memory */
+	if (dpmi_lock_linear_region_base((void near *)handler, size)) {
+		return NULL;
+	}
+
+	irq = (struct irq_handle *) malloc(sizeof(struct irq_handle));
+	irq->c_handler = handler;
+	irq->handler_size = size;
+	irq->prev_vect = old_vect;
+	irq->int_num = intno;
+	irq->irq_num = irqno;
+	irq->pic_base = irqno < 8 ? PIC1_BASE : PIC2_BASE;
+
+	if (dpmi_lock_linear_region_base(irq, sizeof(struct irq_handle))) {
+		free(irq);
+		dpmi_unlock_linear_region_base((void near *)handler, size);
+		return NULL;
+	}
+
+	_dos_setvect(irq->int_num, irq->c_handler);
+
+	irq->pic_mask = irq_state(irq);
+	return irq;
+}
+
+void irq_unhook(struct irq_handle *irq)
+{
+	if (!irq)
+		return;
+
+	/* Restore the interrupt vector */
+	irq_disable(irq);
+	_dos_setvect(irq->int_num, irq->prev_vect);
+
+	/* Unlock the interrupt handler */
+	dpmi_unlock_linear_region_base((void near *)irq->c_handler, irq->handler_size);
+
+	/* Unlock the irq_handle structure */
+	dpmi_unlock_linear_region_base(irq, sizeof(struct irq_handle));
+
+	/* If IRQ was enabled before we hooked, restore enabled state */
+	if (irq->pic_mask)
+		irq_enable(irq);
+	else
+		irq_disable(irq);
+
+	free(irq);
+}
+#endif
+
+
+/*---------------------------------------------- IRQ detection mechanism -----*/
+static struct irq_handle *__irqs[16];
+static int (*__irq_confirm) (int irqno);
+static volatile unsigned int __irq_mask;
+static volatile unsigned int __irq_count[16];
+
+#define DECLARE_IRQ_HANDLER(irqno)							\
+static void INTERRUPT_ATTRIBUTES __irq##irqno##_handler ()						\
+{															\
+  if (irq_check (__irqs [irqno]) && __irq_confirm (irqno))	\
+  {															\
+    __irq_count [irqno]++;									\
+    __irq_mask |= (1 << irqno);								\
+  }															\
+  irq_ack (__irqs [irqno]);									\
+}															\
+static void __irq##irqno##_end(void)								\
```

Comment:
> My main concern here is: is a miscompilation, so to speak, possible here? E.g. can the compiler not reorder things?

One of the comments [here](https://stackoverflow.com/questions/42111493/protected-hardware-interrupt-handler-stuck-djgpp) implies that it's theoretically possible, but the same approach is already used elsewhere in the original MikMod code, as well as in [OpenWatcom's documentation](https://open-watcom.github.io/open-watcom-v2-wikidocs/pguide.html#How_Can_I_Use_the_Mouse_Interrupt_L0x33R_with_DOSD4GW_) and notable libraries like Allegro, so it should be okay in practice.

Reply:
It is safe with multiple `xxx` and `xxx_end` pairs?

Reply:
> It is safe with multiple `xxx` and `xxx_end` pairs?

It seems to be okay in Allegro, so I'd assume so.

Reply:
I've pushed an additional commit that uses [``__attribute__((no_reorder))``](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-no_005freorder-function-attribute) in GCC builds, which should prevent the compiler from reordering the relevant functions.

Reply:
That seems to be supported by new gcc versions, I wonder which.. (my old djgpp gcc3.4 surely doesn't support it.)

We can, though, remove the interrupt code and support only sb >= 2.0 (like quake) with auto-dma only.